### PR TITLE
Upgraded module to hapi v17

### PR DESCRIPTION
**Pull request by mrlannigan**
**Breaking change**
- Updates `validateFunc` function signature, see Readme.

This includes updates to all tests, example, and documentation.

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -16,1680 +15,1339 @@ const internals = {};
...
 
         const server = new Hapi.Server();
-        server.connection();
-        server.register(require('../'), (err) => {
+        server.register(require('../'));
+
+        expect(() => {
 
-            expect(err).to.not.exist();
+            server.auth.strategy('session', 'cookie', { password: new Buffer('foobar') });
```
Comment (vinicius0026) [inline test/index.js]: `new Buffer` [is deprecated](https://nodejs.org/docs/latest-v8.x/api/buffer.html#buffer_new_buffer_string_encoding) since node v6. Should be replaced by `Buffer.from`.

Comment (mrlannigan) [reply test/index.js]: Good catch! Updated.

File: test/index.js [hunk 2]
```diff
@@ -16,1680 +15,1339 @@ const internals = {};
...
 
 describe('scheme', () => {
 
-    it('fails with no plugin options', (done) => {
+    it('fails with no plugin options', () => {
 
         const server = new Hapi.Server();
-        server.connection();
-        server.register(require('../'), (err) => {
+        server.register(require('../'));
```
Comment (vinicius0026) [inline test/index.js]: Shouldn't you `await` on `server.register` as per the new [docs](https://github.com/hapijs/hapi/blob/master/API.md#server.register())

File: test/index.js [hunk 3]
```diff
@@ -16,1680 +15,1339 @@ const internals = {};
...
-            expect(() => {
+        expect(res.result).to.equal('valid');
+        const header = res.headers['set-cookie'];
+        expect(header.length).to.equal(1);
+        expect(header[0]).to.contain('Max-Age=60');
+        const cookie = header[0].match(/(?:[^\x00-\x20\(\)<>@\,;\:\\"\/\[\]\?\=\{\}\x7F]+)\s*=\s*(?:([^\x00-\x20\"\,\;\\\x7F]*))/);
 
-                server.auth.strategy('default', 'cookie', true, { validateFunc: 'not a function' });
-            }).to.throw(Error);
+        /* eslint-disable hapi/no-shadow-relaxed */
```
Comment (vinicius0026) [inline test/index.js]: I guess this flag is no longer necessary

File: test/index.js (hunk 2 above)
Comment (nlf) [reply test/index.js]: yes, you should

File: test/index.js (hunk 3 above)
Comment (nlf) [reply test/index.js]: yeah, if you need to turn off linting in your tests your tests are broken - it looks like you've fixed the error this was meant to circumvent anyway though

File: README.md [hunk 4]
```diff
@@ -130,75 +129,74 @@ const login = function (request, reply) {
...
+    request.cookieAuth.set({ sid });
 
-        request.cookieAuth.set({ sid: sid });
-        return reply.redirect('/');
-    });
+    return h.redirect('/');
 };
 
-const logout = function (request, reply) {
+const logout = function (request, h) {
```
Comment (nlf) [inline README.md]: fat arrow functions

File: README.md [hunk 5]
```diff
@@ -130,75 +129,74 @@ const login = function (request, reply) {
...
 
     request.cookieAuth.clear();
-    return reply.redirect('/');
+    return h.redirect('/');
 };
 
-const server = new Hapi.Server();
-server.connection({ port: 8000 });
+const server = new Hapi.Server({ port: 8000 });
```
Comment (nlf) [inline README.md]: `const server = Hapi.server({ port: 8000 });`

File: README.md [hunk 6]
```diff
@@ -130,75 +129,74 @@ const login = function (request, reply) {
...
-    server.auth.strategy('session', 'cookie', true, {
+    server.auth.strategy('session', 'cookie', {
         password: 'password-should-be-32-characters',
         cookie: 'sid-example',
         redirectTo: '/login',
         isSecure: false,
-        validateFunc: function (request, session, callback) {
-
-            cache.get(session.sid, (err, cached) => {
+        validateFunc: async function (request, session) {
```
Comment (nlf) [inline README.md]: fat arrow

File: README.md [hunk 7]
```diff
@@ -130,75 +129,74 @@ const login = function (request, reply) {
...
-    });
-});
+    console.log('Server ready');
+};
+
+try {
+    exports.start();
+}
+catch (err) {
+    console.log(err.stack);
```
Comment (nlf) [inline README.md]: i feel like there should be a `process.exit(1);` here, as-is this would log an error but exit with a non-error exit code, that seems bad

File: example/index.js [hunk 8]
```diff
@@ -12,19 +12,19 @@ const users = {
     }
 };
 
-const home = function (request, reply) {
+const home = function (request, h) {
```
Comment (nlf) [inline example/index.js]: fat arrow

File: example/index.js [hunk 9]
```diff
@@ -50,74 +50,73 @@ const login = function (request, reply) {
...
+    request.cookieAuth.set({ sid });
 
-        request.cookieAuth.set({ sid: sid });
-        return reply.redirect('/');
-    });
+    return h.redirect('/');
 };
 
-const logout = function (request, reply) {
+const logout = function (request, h) {
```
Comment (nlf) [inline example/index.js]: fat arrow

File: example/index.js [hunk 10]
```diff
@@ -50,74 +50,73 @@ const login = function (request, reply) {
...
 
     request.cookieAuth.clear();
-    return reply.redirect('/');
+    return h.redirect('/');
 };
 
-const server = new Hapi.Server();
-server.connection({ port: 8000 });
+const server = new Hapi.Server({ port: 8000 });
```
Comment (nlf) [inline example/index.js]: `const server = Hapi.server({ port: 8000 });`

File: example/index.js [hunk 11]
```diff
@@ -50,74 +50,73 @@ const login = function (request, reply) {
...
     server.app.cache = cache;
 
-    server.auth.strategy('session', 'cookie', true, {
+    server.auth.strategy('session', 'cookie', {
         password: 'password-should-be-32-characters',
         cookie: 'sid-example',
         redirectTo: '/login',
         isSecure: false,
-        validateFunc: function (request, session, callback) {
+        validateFunc: async function (request, session) {
```
Comment (nlf) [inline example/index.js]: fat arrow

File: test/index.js [hunk 12]
```diff
@@ -16,1680 +15,1339 @@ const internals = {};
...
+// Helpers
+const Helpers = require('./helpers');
 
 
 describe('scheme', () => {
 
-    it('fails with no plugin options', (done) => {
+    it('fails with no plugin options', () => {
 
         const server = new Hapi.Server();
```
Comment (nlf) [inline test/index.js]: `const server = Hapi.server()`

File: example/index.js [hunk 13]
```diff
@@ -50,74 +50,73 @@ const login = function (request, reply) {
...
-    });
-});
+    console.log('Server ready');
+};
+
+try {
+    exports.start();
+}
+catch (err) {
+    console.log(err.stack);
```
Comment (nlf) [inline example/index.js]: again, `process.exit(1);`

File: README.md (hunk 7 above)
Comment (Marsup) [reply README.md]: I doubt the catch would ever be triggered since node doesn't have top level async/await.

Comment (nlf) [reply README.md]: good point, it would be a promise rejection, so `exports.start().catch((err) => {});` is more appropriate

Comment (mrlannigan) [reply README.md]: So what do you all think about wrapping this in a self-executing async function?

Comment (mrlannigan) [reply README.md]: Also, I didn't really want to make too large of a change in this PR aside from making it compatible with hapi17, but I'm not a huge fan of the fact the whole `example/index.js` is in the readme. I feel like I could just keep the example updated and point to it from here.

File: example/index.js (hunk 10 above)
Comment (wswoodruff) [reply example/index.js]: Are we not using `new` anymore? I thought I saw that somewhere.. maybe I was reading the wrong docs version!

Comment (mrlannigan) [reply example/index.js]: https://hapijs.com/api/17.0.1#-serveroptions
compared to:
https://hapijs.com/api/16.6.2#new-serveroptions

And it looks like this is where it was changed to a factory method:
https://github.com/hapijs/hapi/commit/f35b8c5c8a1958f3a3eb9462b4a1ad537a28fe57#diff-c945a46d13b34fcaff544d966cffcabaR23

Comment (wswoodruff) [reply example/index.js]: Cool, thanks for the research!

File: lib/index.js [hunk 14]
```diff
@@ -73,15 +70,13 @@ internals.implementation = function (server, options) {
 
     server.state(settings.cookie, cookieOptions);
 
-    const decoration = function (request) {
+    const decoration = (request) => {
 
-        const CookieAuth = function () {
+        class CookieAuth {
```
Comment (Marsup) [inline lib/index.js]: Probably the class can be extracted and `request` attached to it through the constructor.

Comment (mrlannigan) [reply lib/index.js]: Done, also had to attach the settings.

**Comment by toddhickerson** [comment]
Any ETA on when this will be completed, or should someone else provide a new PR?

File: package.json [hunk 15]
```diff
@@ -12,20 +12,21 @@
     "session"
   ],
   "engines": {
-    "node": ">=4.x.x"
+    "node": ">=8.9.x"
   },
   "dependencies": {
     "boom": "3.x.x",
```
Comment (Marsup) [inline package.json]: I think boom needs a well-deserved bump.

**Comment by Marsup** [comment]
@toddhickerson Why a new PR ?

**Comment by aknuds1** [comment]
@nlf Are there any remaining issues now?

**Comment by toddhickerson** [comment]
@Marsup Not sure how the whole process works, but what is remaining to get this accepted? This is the last item holding me back from hapi v17 on two projects.

File: package.json (hunk 15 above)
Comment (mrlannigan) [reply package.json]: Good point, done.

File: README.md [hunk 16]
```diff
@@ -42,20 +42,19 @@ The `'cookie`' scheme takes the following options:
   query component of the `redirectTo` URI using the parameter name `'next'`. Set to a string to use
   a different parameter name. Defaults to `false`.
 - `redirectOnTry` - if `false` and route authentication mode is `'try'`, authentication errors will
-  not trigger a redirection. Requires **hapi** version 6.2.0 or newer. Defaults to `true`;
-- `validateFunc` - an optional session validation function used to validate the content of the
+  not trigger a redirection. Defaults to `true`;
+- `await validateFunc` - an optional session validation function used to validate the content of the
```
Comment (hueniverse) [inline README.md]: This should be `async` here as it is the method provided by the user, not invoked by the user.

**Comment by aknuds1** [comment]
I've tested this PR now and it works AFAICT.

File: README.md (hunk 16 above)
Comment (mrlannigan) [reply README.md]: Makes sense, updated.

File: README.md [hunk 17]
```diff
@@ -42,20 +42,19 @@ The `'cookie`' scheme takes the following options:
   query component of the `redirectTo` URI using the parameter name `'next'`. Set to a string to use
   a different parameter name. Defaults to `false`.
 - `redirectOnTry` - if `false` and route authentication mode is `'try'`, authentication errors will
-  not trigger a redirection. Requires **hapi** version 6.2.0 or newer. Defaults to `true`;
-- `validateFunc` - an optional session validation function used to validate the content of the
+  not trigger a redirection. Defaults to `true`;
+- `async validateFunc` - an optional session validation function used to validate the content of the
```
Comment (mtharrison) [inline README.md]: Do you think `async` should be here? It could confuse people, thinking that async is part of the property name? In hapi-auth-basic I put it as part of the signature: https://github.com/hapijs/hapi-auth-basic#hapi-auth-basic

File: README.md [hunk 18]
```diff
@@ -130,75 +129,73 @@ const login = function (request, reply) {
...
 
-    server.start(() => {
+    await server.start();
 
-        console.log('Server ready');
-    });
+    console.log('Server ready');
+};
+
+exports.start().catch((err) => {
```
Comment (mtharrison) [inline README.md]: https://github.com/hapijs/hapi/pull/3684#issuecomment-345908229

File: test/helpers.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+                request.cookieAuth.set({ user: request.params.user });
+                return h.response(request.params.user);
+            }
+        }
+    });
+
+    server.route({
+        method: 'GET', path: '/resource', handler: function (request, h) {
+
+            return h.response('resource');
```
Comment (mtharrison) [inline test/helpers.js]: Can just return it?

File: test/helpers.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+exports.loginWithResourceEndpoint = (server) => {
+
+    server.route({
+        method: 'GET', path: '/login/{user}',
+        config: {
+            auth: { mode: 'try' },
+            handler: function (request, h) {
+
+                request.cookieAuth.set({ user: request.params.user });
+                return h.response(request.params.user);
```
Comment (mtharrison) [inline test/helpers.js]: Can just return it?

File: README.md (hunk 17 above)
Comment (hueniverse) [reply README.md]: In hapi docs its in the description signature.

File: lib/index.js [hunk 21]
```diff
@@ -156,38 +156,50 @@ internals.implementation = function (server, options) {
...
 
-                    if (err ||
-                        !isValid) {
+                try {
+                    const result = await settings.validateFunc(request, session);
 
-                        if (settings.clearInvalid) {
-                            reply.unstate(settings.cookie);
-                        }
+                    Hoek.assert(typeof result === 'object', 'Invalid return from validateFunc');
```
Comment (mtharrison) [inline lib/index.js]: Here you're validating the user's own code. It feels a bit out of scope, they should have tests.

Comment (aknuds1) [reply lib/index.js]: Personally I find it useful that the framework verifies the results of user provided hooks for ease of debugging, unless the benefits are outweighed by performance concerns.

Comment (mtharrison) [reply lib/index.js]: There's no harm in it I guess. I don't think there's any perf concerns with a couple of asserts.

**Comment by toddhickerson** [comment]
@mrlannigan I appreciate your work on this and am anxious to use it with Hapi v17. Are there any remaining issues to be addressed?

**Comment by aknuds1** [comment]
I'm also anxious to see this being integrated, I'm already using it (via [my namespaced package](https://www.npmjs.com/package/@arve.knudsen/hapi-auth-cookie)) :)

**Comment by BigWillie** [comment]
Will this be accepted? Or should I go with the namespaced package: https://www.npmjs.com/package/@arve.knudsen/hapi-auth-cookie

**Comment by nlf** [comment]
@mrlannigan i'm going to make you the new maintainer of this module and let you merge and publish this at will. what's your npm username so we can get you permissions there?

**Comment by nlf** [comment]
also, you have a pending invitation to the hapijs org, iirc you can go to https://github.com/hapijs to accept it

**Comment by mrlannigan** [comment]
@nlf Thank you!

@toddhickerson I am going to address @mtharrison changes then work on a release.

**Comment by nlf** [comment]
@mrlannigan what's your npm username?

File: test/helpers.js (hunk 20 above)
Comment (mrlannigan) [reply test/helpers.js]: You can.

**Comment by mrlannigan** [comment]
Sorry @nlf, it's the same as my github username: [mrlannigan](https://www.npmjs.com/~mrlannigan)

**Comment by nlf** [comment]
No worries, I just gave you publish permissions there. Module's all yours!

**Comment by mtharrison** [comment]
It was merged without any approvals?

**Comment by nlf** [comment]
It’s his module now and his pull request, it’s on him to maintain as he sees fit

**Comment by mrlannigan** [comment]
@mtharrison I felt like I had addressed all the blocking changes and next time I will definitely either explicitly ask for the disapproval to be reevaluated or I will dismiss it with an explanation.
